Feature/smhe 2121 overwrite functie shipmentfiles

natasja-n opened the pull request:
SMHE-2121 Overwrite functie in het inladen van shipment files
Add overwrite option for device

https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/253/

natasja-n committed:
SMHE-2121 sf overwrite
w.i.p. test

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
SMHE-2121 sf overwrite
w.i.p. test

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
SMHE-2121 sf overwrite
w.i.p. test

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
SMHE-2121 sf overwrite
w.i.p. test

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
SMHE-2121 sf overwrite
add overwrite

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
Merge branch 'development' into feature/SMHE-2121_Overwrite_functie_shipmentfiles

natasja-n committed:
SMHE-2121 SF overwrite
clean up

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

harrymiddelburgetall commented on osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/installation-ws-smartmetering.xsd:
```diff
@@ -12,6 +12,7 @@
       <xsd:sequence>
         <xsd:element name="Device" type="tns:Device" />
         <xsd:element name="DeviceModel" type="tns:DeviceModel" />
+        <xsd:element name="overwrite" type="xsd:boolean" />
```
default false?

natasja-n replied:
default="false" is added.

harrymiddelburgetall commented on osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java:
```diff
@@ -69,23 +70,48 @@ public InstallationService() {
+      if (addSmartMeterRequest.getOverwrite()) {
+        smartMeter.setVersion(existingSmartMeter.get().getVersion());
+        smartMeter.setId(existingSmartMeter.get().getId());
+      } else {
+        LOGGER.error(
+            "ERROR: SmartMeter with device identification {} already exists and overwrite is not enabled",
+            messageMetadata.getDeviceIdentification());
+      }
```
This is already covered by the _validateSmartMeterDoesNotExist_ method, right?

natasja-n replied:
Right, removed duplicate code and updated method name to checkIfSmartMeterExistsAndOverwriteAllowed

harrymiddelburgetall commented on osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java:
```diff
@@ -69,23 +70,48 @@ public InstallationService() {
+            .getMapperFacade()
+            .map(addSmartMeterRequest.getDevice(), SmartMeteringDeviceDto.class);
+    requestDto.setOverwrite(addSmartMeterRequest.getOverwrite());
+    this.osgpCoreRequestMessageSender.send(requestDto, messageMetadata); // en dan??
```
_en dan??_
Is there an open end to the logic?

natasja-n replied:
sorry, removed comment.

harrymiddelburgetall commented on osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationServiceTest.java:
```diff
@@ -121,18 +125,69 @@ public void addMeterProtocolInfoNotFound() {
+
+  @Test
+  void testUpdateSmartMeter() throws FunctionalException {
+    final SmartMeter existingSmartMeter = new SmartMeter();
+    existingSmartMeter.setId(1L);
+    existingSmartMeter.setVersion(2L);
+
+    when(this.smartMeteringDeviceRepository.findByDeviceIdentification(DEVICE_IDENTIFICATION))
+        .thenReturn(existingSmartMeter);
+    when(this.addSmartMeterRequest.getOverwrite()).thenReturn(true);
+
+    doReturn(existingSmartMeter)
+        .when(this.smartMeterService)
+        .convertSmartMeter(this.smartMeteringDevice);
+
+    when(this.protocolInfoRepository.findByProtocolAndProtocolVersionAndProtocolVariant(
+            this.smartMeteringDevice.getProtocolName(),
+            this.smartMeteringDevice.getProtocolVersion(),
+            this.smartMeteringDevice.getProtocolVariant()))
+        .thenReturn(this.protocolInfo);
```
These lines of codes are equals with the next test except for the overwrite boolean. Did you consider a parameterized test?

natasja-n replied:
I use the ParameterizedTest with overwrite boolean in the smart-meter-headend pr, but in some unexplainable way not in this one. They are all updated and merged now.

harrymiddelburgetall commented on osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java:
```diff
@@ -64,10 +65,28 @@ void testNonExistingSmartMeter() {
   }
 
+  @Test
+  void testNonExistingSmartMeterWithOverwriteIsTrue() throws FunctionalException {
```
This is a test with an existing smart meter. Confusing method name.

natasja-n replied:
Tests are updated.

harrymiddelburgetall commented on osgp/platform/osgp-adapter-domain-smartmetering/src/test/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/SmartMeterServiceTest.java:
```diff
@@ -64,10 +65,28 @@ void testNonExistingSmartMeter() {
 
+  @Test
+  void testNonExistingSmartMeterWithOverwriteIsTrue() throws FunctionalException {
+    final boolean overwrite = true;
```
I'm missing a test with an existing smart meter and a overwrite false parameter

natasja-n replied:
Tests are updated.

harrymiddelburgetall commented on osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java:
```diff
@@ -144,6 +147,92 @@ void addMeterNoDeviceIdentification() {
+  }
+
+  @Test
+  void testAddMeterWithNoExistingDeviceIsNullAndOverwriteIsFalse() throws FunctionalException {
```
This test only differs from the previous (testAddMeterWithNoExistingDeviceIsNullAndOverwriteIsTrue) except for the overwrite boolean. Please use a parameterized test for this.

natasja-n replied:
Right, tests are updated.

harrymiddelburgetall commented on osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/InstallationServiceTest.java:
```diff
@@ -144,6 +147,92 @@ void addMeterNoDeviceIdentification() {
+  }
+
+  @Test
+  void testAddMeterWithExistingDeviceAndOverwriteIsTrue() throws FunctionalException {
```
If you add an boolean to the parameterized test suggested above  to indicate existing device or not you can merge the three tests.

natasja-n replied:
Tests are updated.

harrymiddelburgetall reviewed (commented):
Complex piece of functionality you implemented here. My comments mainly cover the tests

harrymiddelburgetall commented on osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java:
```diff
@@ -69,23 +70,48 @@ public InstallationService() {
     final SmartMeter smartMeter = this.smartMeterService.convertSmartMeter(smartMeteringDevice);
+    if (existingSmartMeter.isPresent()) {
+      if (addSmartMeterRequest.getOverwrite()) {
+        smartMeter.setVersion(existingSmartMeter.get().getVersion());
```
Sonar states that the id field you added to the smartmeter entity is already existing in the AbstractEntity superclass. You can potentialy  avoid this construction with the hibernate merge approach.

natasja-n replied:
Sonar issues are resolved. With the id and version the SmartMeter object is under control.

harrymiddelburgetall reviewed (changes requested):
One additional comment after looking at sonar

natasja-n committed:
SMHE-2121 SF overwrite
review comments: user @ParameterizedTest for overwrite boolean, remove comment and duplicate check, set overwrite default false

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
Merge branch 'development' into feature/SMHE-2121_Overwrite_functie_shipmentfiles

natasja-n committed:
SMHE-2121 SF overwrite
resolve sonar issues, remove unused method

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>

natasja-n committed:
SMHE-2121 SF overwrite
update unit tests

Signed-off-by: Natasja Nortier <natasja.nortier@alliander.com>
